packages/contracts/contracts/L1/ZKMerkleTrie.sol

```diff
@@ -99,7 +99,7 @@ contract ZKMerkleTrie is IZKMerkleTrie, ZKTrieHasher {
                 );
             } else if (currentNode.nodeType == NodeReader.NodeType.LEAF) {
                 require(!exists && !empty, "ZKMerkleTrie: duplicated terminal node");
-                exists = true;
+                exists = currentNode.nodeKey == _key;
```

Comment:
If it doesn't exist, you can break the loop early. Plus, thanks to adding this line, we can safely remove the if condition below `if (currentNode.keyPreimage...`)

Reply:
I will process it as a require to maintain uniformity with other codes.
Shouldn't we think more about removing the verification part with keyPrimage? If that part remains, it can also be verified in the event of a hash collision.